nixos/tests/k3s/default.nix

```
@@ -1,9 +1,21 @@
 { system ? builtins.currentSystem
 , pkgs ? import ../../.. { inherit system; }
+, lib ? pkgs.lib
 }:
+let
+  allK3s = {
+    inherit (pkgs)
+      k3s
+      k3s_1_23
+      k3s_1_24
+      k3s_1_25
+      k3s_1_26
+      ;
+  };
```

```suggestion
  allK3s =
    let k3sDir = builtins.readDir ../../../pkgs/applications/networking/cluster/k3s; in
      lib.attrsets.mapAttrs (name: _value: lib.attrsets.getAttrFromPath [ "k3s_${name}" ] pkgs ) k3sDir;
```

And k3s (with no version) doesn't have to be there because it could trigger an extra build?
I haven't checked, just wondering. If version is already listed, it would be tested anyway?

---

Now I wonder if from performance POV this is a bad idea. =/

---

Because of possible performance issues, nevermind this.

---

Really listing files in a directory is a performance penalty? We're not rendering a Pixar movie here... 😅

---

> Really listing files in a directory is a performance penalty? We're not rendering a Pixar movie here... sweat_smile

You liked idea, right? Me too.

---

Maybe I can add the suggestion and, if some performance penalty _really_ happens, then we can always roll back to the hardcoded list. WDYT?

---

@SuperSandro2000 Is it okay doing dynamic evaluation, like this:
```
  allK3s =
    let k3sDir = builtins.readDir ../../../pkgs/applications/networking/cluster/k3s; in
    lib.attrsets.mapAttrs' (name: _value: {
      name = "k3s_${name}";
      value = lib.attrsets.getAttrFromPath [ "k3s_${name}" ] pkgs;
    }) k3sDir;
```
?

---

We could also `allK3s = lib.filterAttrs (n: _: lib.strings.hasPrefix "k3s_" n) pkgs;` if we wanted, right?

It does feel like avoiding hardcoding this list here would be nice. Especially once the update script can add a new major version, it would be nice for it to get tested without having to remember to also update this file... though I guess it will still require updating `all-packages.nix` for new major versions regardless.

---

> We could also `allK3s = lib.filterAttrs (n: _: lib.strings.hasPrefix "k3s_" n) pkgs;` if we wanted, right?

Seems good.

---

@yajo Feel free to implement `allK3s` as euank suggested.